### PR TITLE
Add Nuxt.js recipe

### LINKED ISSUE #2063: NuxtJS/VueJS recipe for ava
Hello, I'm having an error on my component testing when there is a vue mixin on it

```
  Uncaught exception in test/specs/Logo.spec.js

  /home/sprguillen/Workspace/unit-test/components/Logo.vue:7

   6:     <div class="Triangle Triangle--four" />
   7:   </div>                                   
   8: </template>                                

  Error: Cannot find module '~/mixins/form-mixin'
```

Here is my configuration:

setup.js

```
require('browser-env')()
require('jsdom-global')()
const hooks = require('require-extension-hooks')
const Vue = require('vue')

Vue.config.productionTip = false

window.Date = Date;

hooks('vue').plugin('vue').push();
hooks(['vue', 'js']).plugin('babel').push();
```

ava.config.js

```
export default {
  require: [
    "./test/helpers/setup.js"
  ]
}
```
logo.spec.js

```
import { mount } from '@vue/test-utils'
import test from 'ava'
import Logo from '../../components/Logo.vue'

test('is a Vue instance', t => {
  const wrapper = mount(Logo)
  t.is(wrapper.isVueInstance(), true)
})
```

Logo.vue

```
<template>
  <div class="VueToNuxtLogo">
    <div class="Triangle Triangle--two" />
    <div class="Triangle Triangle--one" />
    <div class="Triangle Triangle--three" />
    <div class="Triangle Triangle--four" />
  </div>
</template>

<script>
import FormMixin from '~/mixins/form-mixin'
export default {
  mixins: [FormMixin]
}
</script>


<style>
.VueToNuxtLogo {
  display: inline-block;
  animation: turn 2s linear forwards 1s;
  transform: rotateX(180deg);
  position: relative;
  overflow: hidden;
  height: 180px;
  width: 245px;
}

.Triangle {
  position: absolute;
  top: 0;
  left: 0;
  width: 0;
  height: 0;
}

.Triangle--one {
  border-left: 105px solid transparent;
  border-right: 105px solid transparent;
  border-bottom: 180px solid #41b883;
}

.Triangle--two {
  top: 30px;
  left: 35px;
  animation: goright 0.5s linear forwards 3.5s;
  border-left: 87.5px solid transparent;
  border-right: 87.5px solid transparent;
  border-bottom: 150px solid #3b8070;
}

.Triangle--three {
  top: 60px;
  left: 35px;
  animation: goright 0.5s linear forwards 3.5s;
  border-left: 70px solid transparent;
  border-right: 70px solid transparent;
  border-bottom: 120px solid #35495e;
}

.Triangle--four {
  top: 120px;
  left: 70px;
  animation: godown 0.5s linear forwards 3s;
  border-left: 35px solid transparent;
  border-right: 35px solid transparent;
  border-bottom: 60px solid #fff;
}

@keyframes turn {
  100% {
    transform: rotateX(0deg);
  }
}

@keyframes godown {
  100% {
    top: 180px;
  }
}

@keyframes goright {
  100% {
    left: 70px;
  }
}
</style>

```

form-mixin.js (Take note, I shortened the actual code but regardless the point is, it's not working. This is a working mixin).

```
export default {
  data() {
    return {
      errors: {}
    }
  }
}
```

I just want ava or vue test-utils to recognize namespaces either using `~/mixins` or `@/mixins` and yes, mixins directory exists and it's inside the root folder similar to components, pages, layouts, test etc.

Finally here is my package.json file:

```
{
  "name": "unit-test",
  "version": "1.0.0",
  "description": "My amazing Nuxt.js project",
  "author": "The Handsome Boy",
  "private": true,
  "scripts": {
    "dev": "nuxt",
    "build": "nuxt build",
    "start": "nuxt start",
    "generate": "nuxt generate",
    "test": "ava"
  },
  "dependencies": {
    "cross-env": "^5.2.0",
    "nuxt": "^2.3.4",
    "@nuxtjs/pwa": "^2.6.0"
  },
  "devDependencies": {
    "@vue/test-utils": "^1.0.0-beta.27",
    "ava": "^1.0.1",
    "browser-env": "^3.2.5",
    "jsdom-global": "^3.0.2",
    "nodemon": "^1.18.9",
    "require-extension-hooks": "^0.3.3",
    "require-extension-hooks-babel": "^0.1.1",
    "require-extension-hooks-vue": "^2.0.0"
  }
}
```

### PR DESCRIPTION
**Pull request by vinayakkulkarni**
Fixes #2063 

- Boilerplate for new user of `Nuxt` as well as `Ava` could be found [here](https://github.com/vinayakkulkarni/nuxt-ava-e2e-unit-testing)

### CODE REVIEW

**Comment by vinayakkulkarni** [comment]
bump

File: docs/recipes/nuxt.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+First set up a helper to configure the environment to transpile `.vue` files and run in a browser like environment.
+
+**`package.json`:**
+
+```json
+{
+	"ava": {
+		"require": [
+			"./test/helpers/register.js",
+			"./test/helpers/setup.js"
```
Comment (novemberborn) [inline docs/recipes/nuxt.md]: With AVA v2, `test/helpers` files are no longer automatically considered to be helpers. Could you change these to `test/_setup.js`?

File: docs/recipes/nuxt.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+```js
+// ./test/helpers/register.js
+
+require('@babel/register')({
+	// These patterns are relative to the project root directory (where the `package.json` file lives):
+	ignore: [
+		'node_modules/*',
+		'test/*'
+	]
+});
```
Comment (novemberborn) [inline docs/recipes/nuxt.md]: Why not directly require `@babel/register`?

File: docs/recipes/nuxt.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+// Setup Vue.js to remove production tip
+Vue.config.productionTip = false;
+
+// https://github.com/nuxt/create-nuxt-app/issues/180
+window.Date = global.Date = Date;
+
+// Setup `.vue` files to be processed by `require-extension-hooks-vue`
+hooks('vue').plugin('vue').push();
+// Setup `.vue` and `.js` files to be processed by `require-extension-hooks-babel`
+hooks(['vue', 'js']).exclude(({filename}) => filename.match(/\/node_modules\//)).plugin('babel').push();
```
Comment (novemberborn) [inline docs/recipes/nuxt.md]: Why this *and* `@babel/register`?

File: docs/recipes/nuxt.md [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+```
+
+**`.babelrc`:**
```
Comment (novemberborn) [inline docs/recipes/nuxt.md]: I believe Babel is recommending the config goes into `babel.config.js` files these days. Why does this require Babel at all?

File: docs/recipes/nuxt.md [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+			"plugins": [
+				[
+					"module-resolver",
+					{
+						"root": [
+							"."
+						],
+						"alias": {
+							"@": ".",
+							"~": "."
```
Comment (novemberborn) [inline docs/recipes/nuxt.md]: Are these typical aliases for Vue projects? It seems irrelevant for this recipe.

File: docs/recipes/nuxt.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+const localVue = createLocalVue();
+
+test.beforeEach(() => {
+	wrapper = mount(Index, {
+		localVue,
+	});
+});
+
+test('is a Vue instance', t => {
+	t.is(wrapper.isVueInstance(), true);
```
Comment (novemberborn) [inline docs/recipes/nuxt.md]: Could you use `t.true()`?

Comment (vinayakkulkarni) [reply docs/recipes/nuxt.md]: - [x] Done in 644b2bb

File: docs/recipes/nuxt.md (hunk 5 above)
Comment (vinayakkulkarni) [reply docs/recipes/nuxt.md]: Nuxt also has aliases which are helpful when writing `Vue` components in Nuxt apps.

Ref: https://nuxtjs.org/guide/directory-structure/#aliases

File: docs/recipes/nuxt.md (hunk 1 above)
Comment (vinayakkulkarni) [reply docs/recipes/nuxt.md]: - [x] Done in aed503c

File: docs/recipes/nuxt.md (hunk 2 above)
Comment (vinayakkulkarni) [reply docs/recipes/nuxt.md]: - [x] Done in fb6fcc8

File: docs/recipes/nuxt.md (hunk 3 above)
Comment (vinayakkulkarni) [reply docs/recipes/nuxt.md]: You're right, @babel/register & hooks are redundant.

- [x] Fixed in 73621a1

File: docs/recipes/nuxt.md (hunk 4 above)
Comment (vinayakkulkarni) [reply docs/recipes/nuxt.md]: - Mostly for the `module-resolver`

Comment (vinayakkulkarni) [reply docs/recipes/nuxt.md]: - [x] Fixed in `744d635`

File: docs/recipes/nuxt.md (hunk 5 above)
Comment (vinayakkulkarni) [reply docs/recipes/nuxt.md]: > Are these typical aliases for Vue projects? It seems irrelevant for this recipe.

Yes, There can be components deep inside directories and test directory *ideally* also imitate the project structure (snapshots looks really seksy), so having `../../../../../pages/dashboard/sidebar/nav/logout.vue` would be very ugly inside the test which lies inside `/test/pages/dashboard/sidebar/nav/logout.spec.js`. Hence the `alias`!

File: docs/recipes/nuxt.md [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+```
+
+You can find more information about setting up Babel with AVA in the [Babel recipe](babel.md).
+
+## Sample snapshot test
+
+```js
+import {mount, createLocalVue} from '@vue/test-utils';
+import test from 'ava';
+import Index from '@/pages/index.vue';
```
Comment (novemberborn) [inline docs/recipes/nuxt.md]: Given that this is an example, maybe we could simplify it by removing the `@` resolver? It'll be easier to understand the test example.

You could still document how to hook up Babel so it processes `.vue` files.

File: docs/recipes/nuxt.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+let wrapper;
+const localVue = createLocalVue();
+
+test.beforeEach(() => {
+	wrapper = mount(Index, {
```
Comment (novemberborn) [inline docs/recipes/nuxt.md]: You cannot use a shared variable in a `beforeEach()` hook. It works in this example, because the tests are synchronous. But if you had asynchronous tests then the `beforeEach` hooks would run and change the `wrapper` value.

If it can be shared, use `test.before()`. But perhaps better yet is to assign it to the context:

```js
test.beforeEach(t => {
  t.context.wrapper = mount(Index, { localVue })
})

test('is a Vue instance', t => {
  t.true(t.context.wrapper.isVueInstance())
})
```

**Comment by novemberborn** [comment]
Closing due to inactivity.
